File: app/crypto.py
"""Scheme cynderlab.secret.v1 — Python mirror of static/js/crypto.js.

The production server never encrypts or decrypts: all real cryptography runs
in the browser. This mirror exists so the test suite (and gate_e2e.py) can act
as a browser, and so the shared test vectors keep both implementations locked
together. The `cryptography` package is therefore a dev-only dependency and is
imported lazily — the runtime only ever touches the stdlib helpers below.
"""

import base64
import re
import secrets

HKDF_INFO = b"cynderlab.secret.v1"
PBKDF2_ITERATIONS = 310_000
SLUG_RE = re.compile(r"^[A-Za-z0-9_-]{22}$")


def new_slug() -> str:
    return secrets.token_urlsafe(16)


def new_key() -> bytes:
    return secrets.token_bytes(32)


def b64u_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def b64u_decode(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def derive_verifier(passphrase: str, slug: str) -> bytes:
    """Proof-of-passphrase sent to the server. Domain-separated from the AES key
    derivation by the '.verify' salt suffix; the server stores only its sha256."""
    from cryptography.hazmat.primitives import hashes
    from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

    return PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=f"{slug}.verify".encode("utf-8"),
        iterations=PBKDF2_ITERATIONS,
    ).derive(passphrase.encode("utf-8"))


def derive_key(link_key: bytes, slug: str, passphrase: str | None) -> bytes:
    from cryptography.hazmat.primitives import hashes
    from cryptography.hazmat.primitives.kdf.hkdf import HKDF
    from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

    salt = b""
    if passphrase:
        salt = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=slug.encode("utf-8"),
            iterations=PBKDF2_ITERATIONS,
        ).derive(passphrase.encode("utf-8"))
    return HKDF(
        algorithm=hashes.SHA256(), length=32, salt=salt, info=HKDF_INFO
    ).derive(link_key)


def encrypt(plaintext: bytes, aes_key: bytes, slug: str) -> tuple[bytes, bytes]:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    nonce = secrets.token_bytes(12)
    ciphertext = AESGCM(aes_key).encrypt(nonce, plaintext, slug.encode("utf-8"))
    return nonce, ciphertext


def decrypt(nonce: bytes, ciphertext: bytes, aes_key: bytes, slug: str) -> bytes:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM

    return AESGCM(aes_key).decrypt(nonce, ciphertext, slug.encode("utf-8"))

File: static/js/crypto.js
/* Mirror of app/crypto.py — scheme cynderlab.secret.v1.
 * Test vectors (tests/test_crypto.py):
 *   linkKey=bytes 0..31, slug="AAAAAAAAAAAAAAAAAAAAAA", passphrase="hunter2"
 *   => b64u(aesKeyBits) === "aQ9zwdkp5wqhsrCL5-kxi7yy-sKCAfvDrl0DHgKd5KY"
 *   deriveVerifier("hunter2", "AAAAAAAAAAAAAAAAAAAAAA")
 *   => "6cG5RzzGKJuDk7J761se5TuxDPtCpXizEfxGyjn4gpY" */
(function () {
  "use strict";
  const te = new TextEncoder(), td = new TextDecoder();
  const HKDF_INFO = te.encode("cynderlab.secret.v1");
  const PBKDF2_ITERATIONS = 310000;

  function b64uEncode(bytes) {
    let s = "";
    for (const b of bytes) s += String.fromCharCode(b);
    return btoa(s).replaceAll("+", "-").replaceAll("/", "_").replace(/=+$/, "");
  }
  function b64uDecode(str) {
    const s = atob(str.replaceAll("-", "+").replaceAll("_", "/"));
    return Uint8Array.from(s, c => c.charCodeAt(0));
  }
  function randomBytes(n) { return crypto.getRandomValues(new Uint8Array(n)); }
  function newSlug() { return b64uEncode(randomBytes(16)); }

  async function deriveAesKey(linkKey, slug, passphrase, usages) {
    let salt = new Uint8Array(0);
    if (passphrase) {
      const pk = await crypto.subtle.importKey("raw", te.encode(passphrase), "PBKDF2", false, ["deriveBits"]);
      salt = new Uint8Array(await crypto.subtle.deriveBits(
        { name: "PBKDF2", hash: "SHA-256", salt: te.encode(slug), iterations: PBKDF2_ITERATIONS },
        pk, 256));
    }
    const ikm = await crypto.subtle.importKey("raw", linkKey, "HKDF", false, ["deriveKey"]);
    return crypto.subtle.deriveKey(
      { name: "HKDF", hash: "SHA-256", salt, info: HKDF_INFO },
      ikm, { name: "AES-GCM", length: 256 }, false, usages);
  }

  // Proof-of-passphrase for the server gate. Domain-separated from the AES key
  // by the ".verify" salt suffix; the server stores only its sha256.
  async function deriveVerifier(passphrase, slug) {
    const pk = await crypto.subtle.importKey("raw", te.encode(passphrase), "PBKDF2", false, ["deriveBits"]);
    const bits = await crypto.subtle.deriveBits(
      { name: "PBKDF2", hash: "SHA-256", salt: te.encode(slug + ".verify"), iterations: PBKDF2_ITERATIONS },
      pk, 256);
    return b64uEncode(new Uint8Array(bits));
  }

  async function encryptSecret(plaintext, slug, passphrase) {
    const linkKey = randomBytes(32);
    const aesKey = await deriveAesKey(linkKey, slug, passphrase, ["encrypt"]);
    const nonce = randomBytes(12);
    const ct = await crypto.subtle.encrypt(
      { name: "AES-GCM", iv: nonce, additionalData: te.encode(slug) },
      aesKey, te.encode(plaintext));
    return { key: b64uEncode(linkKey), nonce: b64uEncode(nonce),
             ciphertext: b64uEncode(new Uint8Array(ct)) };
  }

  async function decryptSecret(ciphertextB64, nonceB64, keyB64, slug, passphrase) {
    const aesKey = await deriveAesKey(b64uDecode(keyB64), slug, passphrase, ["decrypt"]);
    const pt = await crypto.subtle.decrypt(
      { name: "AES-GCM", iv: b64uDecode(nonceB64), additionalData: te.encode(slug) },
      aesKey, b64uDecode(ciphertextB64));
    return td.decode(pt);
  }

  window.CynderCrypto = { b64uEncode, b64uDecode, randomBytes, newSlug,
                          deriveAesKey, deriveVerifier, encryptSecret, decryptSecret };
})();
